### PR TITLE
Only fire one TARGETTED event per target selected while changing targets.

### LINKED ISSUE #6158: Swift End + Bonecrusher Giants with a Clover gives too much damage
From [reddit](https://www.reddit.com/r/XMage/comments/edthgi/new_release_41v0_with_adventure_cards_new_ai_and/fcvkbvs/):

> Just cast Swift End on 2 Bonecrusher Giants with a Clover out. Would expect to take 8 damage for that ((2+2)x2) but as you can see from the screenshot I took 16.

svoun6R

### PR DESCRIPTION
**Pull request by Dilnu**
This specifically addresses changing the target of a spell or ability on the stack.
This also fixes a bug where TestPlayer doesn't trigger any TARGETTED events.
Fixes #6158

### CODE REVIEW

File: Mage.Tests/src/test/java/org/mage/test/cards/copy/CopySpellTest.java [hunk 1]
```diff
@@ -77,6 +77,29 @@ public void ZadaHedronGrinderBoost() {
...
+        addCard(Zone.HAND, playerA, "Barkshell Blessing");
+
+        castSpell(1, PhaseStep.UPKEEP, playerA, "Barkshell Blessing");
+        setChoice(playerA, "Yes");
+        setChoice(playerA, "Yes");
+        addTarget(playerA, "Grizzly Bears");
+        addTarget(playerA, "Bonecrusher Giant");
+
+        setStopAt(1, PhaseStep.PRECOMBAT_MAIN);
+        execute();
```
Comment (JayDi85) [inline Mage.Tests/src/test/java/org/mage/test/cards/copy/CopySpellTest.java]: Use `setStrictChooseMode` and `assertAllCommandsUsed` all the time in tests (only one exception: AI tests):
```
        setStrictChooseMode(true);
        setStopAt(1, PhaseStep.END_COMBAT);
        execute();
        assertAllCommandsUsed();
```

File: Mage.Tests/src/test/java/org/mage/test/cards/copy/CopySpellTest.java [hunk 2]
```diff
@@ -77,6 +77,29 @@ public void ZadaHedronGrinderBoost() {
         assertAbility(playerB, "Silvercoat Lion", FlyingAbility.getInstance(), false);
     }
 
+    @Test
+    public void BonecrusherGiantChangeTargets() {
+        addCard(Zone.BATTLEFIELD, playerA, "Bonecrusher Giant");
+        addCard(Zone.BATTLEFIELD, playerA, "Grizzly Bears");
+        addCard(Zone.BATTLEFIELD, playerA, "Savannah Lions");
+        addCard(Zone.BATTLEFIELD, playerA, "Plains");
+        addCard(Zone.HAND, playerA, "Barkshell Blessing");
```
Comment (JayDi85) [inline Mage.Tests/src/test/java/org/mage/test/cards/copy/CopySpellTest.java]: Also best practice -- add testing abilities text in test comments:
``` java
// Target creature gets +2/+2 until end of turn.
// Conspire (As you cast this spell, you may tap two untapped creatures you control that share a color with it. When you do, copy it and you may choose a new target for the copy.)
addCard(Zone.HAND, playerA, "Barkshell Blessing");
```

**Comment by JayDi85** [comment]
Look at my [tests from 6158](https://github.com/magefree/mage/issues/6158#issuecomment-581203227). Your code fails on that:
* `BonecrusherGiantChangeTargets_BoneTargetBoth` -- bone must triggers from both spells (original and copied spells targets bone);
* BonecrusherGiantChangeTargets_BoneTargetFirst -- bone must trigger from original spell (original spell target bone, copied spell target another creature).

Please add and check that tests.

**Comment by Dilnu** [comment]
I added the tests you wrote with some minor fixes.

I also found the source of the test errors. It looks like TestPlayer was implemented incorrectly.

**Comment by JayDi85** [comment]
Good finds for test framework bugs

**Pull request merged**